hardware_in_the_loop/software/modules/iocontroller.py

```diff
@@ -0,0 +1,167 @@
+# Extended python imports
+import serial
+import logging
+
+
+class IOController:
+    """High level python object to interface with hardware.
+
+    Used to set analog and digital pins for simulation.
+
+    https://docs.olinelectricmotorsports.com/display/AE/IO+Controller
+    """
+
+    def __init__(self, pin_info_path: str, serial_path: str):
+        # Create logger (all config should already be set by RoadkillHarness)
+        self.log = logging.getLogger(name=__name__)
+
+        self.pin_info = self.read_pin_info(path=pin_info_path)
+        try:
+            self.serial = serial.Serial(port=serial_path, baudrate=115200, timeout=5)
+        except Exception as e:
+            # Couldn't open the specified port; initialize w/o hardware for testing
+            self.log.error(f"Failed to connect to hardware at {pin_info_path}")
+            self.log.error(e)
+            self.serial = None
+
+    def set_state(self, pin: str, value) -> None:
+        """Set the value of an IO pin in the HitL system
+
+        Args:
+            pin (str): The name of the pin to update (e.x. THROTTLE_PEDAL_1)
+
+            value (int or float): The value to set the pin to (e.x. 2.5)
+                - 0 or 1 for digital, volts for analog
+
+        Message format:
+            2 bytes.
+            Byte 1:
+                Bit 0: 1 (indicates a set request)
+                Bits 1-7: Address (int)
+
+            Byte 2: Value
+
+            Value format
+                Digital: 0 or 1 (easy enough)
+                Analog: <5 bit>.<3 bit> floating point value (0 to 31.875 V)
```

Comment:
we should be doing 0-100% for analog signals
i think we also want 2 bytes of precision for this (we need _at least_ 10 bit precision for atmega16m1 DACs, and its not uncommon for MCUs to have 12 bit ADCs)

Reply:
Ah, cool, sounds good to me. In that case, we need to be very clear that `pin_info.csv` should always specify min and max as the voltages of the low and high rails of the DAC being used (not that i really had anything else in mind there, i hadn't thought about it a ton).